src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.cs

```diff
@@ -36,13 +38,57 @@ internal abstract partial class AbstractDiagnosticsTaggerProvider<TTag> : Asynch
     {
         private readonly IDiagnosticService _diagnosticService;
 
+        /// <summary>
+        /// Keep track of the ITextSnapshot for the open Document that was used when diagnostics were
+        /// produced for it.  We need that because the DiagnoticService does not keep track of this
+        /// snapshot (so as to not hold onto a lot of memory), which means when we query it for 
+        /// diagnostics, we don't know how to map the span of the diagnostic to the current snapshot
+        /// we're tagging.
+        /// </summary>
+        private static readonly ConditionalWeakTable<object, ITextSnapshot> _diagnosticIdToTextSnapshot = 
+            new ConditionalWeakTable<object, ITextSnapshot>();
+
         protected AbstractDiagnosticsTaggerProvider(
             IDiagnosticService diagnosticService,
             IForegroundNotificationService notificationService,
             IAsynchronousOperationListener listener)
             : base(listener, notificationService)
         {
             _diagnosticService = diagnosticService;
+            _diagnosticService.DiagnosticsUpdated += OnDiagnosticsUpdated;
+        }
+
+        private void OnDiagnosticsUpdated(object sender, DiagnosticsUpdatedArgs e)
+        {
+            if (e.Solution == null || e.DocumentId == null)
+            {
+                return;
+            }
+
+            if (_diagnosticIdToTextSnapshot.TryGetValue(e.Id, out var snapshot))
+            {
+                return;
+            }
+
+            var document = e.Solution.GetDocument(e.DocumentId);
+
+            // Open documents *should* always have their SourceText available, but we cannot guarantee
+            // (i.e. assert) that they do.  That's because we're not on the UI thread here, so there's
+            // a small risk that between calling .IsOpen the file may then close, which then would
+            // cause TryGetText to fail.  However, that's ok.  In that case, if we do need to tag this
+            // document, we'll just use the current editor snapshot.  If that's the same, then the tags
+            // will be hte same.  If it is different, we'll eventually hear about the new diagnostics 
+            // for it and we'll reach our fixed point.
+            if (document != null && document.IsOpen())
+            {
+                // This should always be fast since the document is open.
+                var sourceText = document.State.GetTextSynchronously(cancellationToken: default);
```

Comment:
Sorry I was missing full context on the previous conversation: should this just be GetTextAsync? In theory still instant, but if we're wrong it's less terrible?

(still OK with this more than TryGetText)

Reply:
Yup, will fix up.